### PR TITLE
NeMo getting started guide

### PR DESCRIPTION
**Pull request by smythp**
## Type of change

Adds a getting started guide for the NeMo Chainguard Image

## Acceptance Criteria / Blockers

Currently this image has CVEs that need to be remediated. I'll keep this as a draft until we can complete this with a 0 or near-0 CVE count. It can still be reviewed, however.

## Testing

The current example can run on a CPU without much problems, but a deep learning environment should also be used to test. I have instructions for setting one up [here](https://github.com/chainguard-images/images/blob/main/images/pytorch-cuda12/TESTING.md#creating-a-deep-learning-vm).

## Notes

I'd be open to adding a finetuning example, but we're waiting on getting some deps into the image. We can wait for more stability / customer requests.

I added a deep learning blurb, I'll put it in the pytorch example after this merges.

I added a widget to play audio files to our shortcodes for this.

I added a details widget that is open by default that I can use for scanner output.

References https://github.com/chainguard-dev/internal/issues/3731

### CODE REVIEW

File: content/chainguard/chainguard-images/getting-started/nemo/index.md [hunk 1]
```diff
@@ -0,0 +1,3013 @@
...
+```
+
+If your host machine does not have attached GPUs and you'd like to run the above on your CPU, omit the `  --gpus all \` line. Since we're using pretrained models to perform text to speech, this example will only take a few minutes using a CPU only. However, other tasks such as model training and finetuning may take significantly longer without connected GPUs.
+
+Note that NeMo collections are large, and initial imports can take up to a minute depending on your environment. The script may appear to hang during that time.
+
+After imports are complete, you should see a large amount of output as NeMo pulls models and works through the steps in the script (tokenizing, generating a spectrogram, generating audio, and writing audio to disk). On completion, the script outputs a `test.wav` file. Because we mounted a volume, this file should now be present in the working directory of your host machine.
+
+```bash
+$ ls
```
[the remark this answers is not in the document]
Comment (erikaheidi) [reply content/chainguard/chainguard-images/getting-started/nemo/index.md:132]: This can look a bit confusing with the command and output together in the same block. I don't think you need to add the `ls` command tbh. If you want to, it's best to have them in separate blocks to optimized rendering:

```shell
ls
```

```
output
```

File: content/chainguard/chainguard-images/getting-started/nemo/index.md [hunk 2]
```diff
@@ -0,0 +1,3013 @@
...
+test.wav  tts.py
+```
+
+Note that we ran the above script as root. This allows us to share the script and output `.wav` file between the host and container. Remember not to run your image as root in a production environment.
```
Comment (erikaheidi) [inline content/chainguard/chainguard-images/getting-started/nemo/index.md:136]: I think this should go right below the command

File: content/chainguard/chainguard-images/getting-started/nemo/index.md [hunk 3]
```diff
@@ -0,0 +1,3013 @@
...
+
+The `test.wav` file should contain audio similar to this output:
+
+{{< audio src="tts-example.wav" caption="Output from the TTS script" >}}
+
+## Advantages of Chainguard Images for Production ML
```
Comment (erikaheidi) [inline content/chainguard/chainguard-images/getting-started/nemo/index.md:142]: I'm not quite sure the getting started guide is the best place for the image comparison, because this data fluctuates a lot and we won't be able to keep it up-to-date. I would advise to remove this or simplify removing the itemized CVEs part. this data should be in the [vul comparison](https://edu.chainguard.dev/chainguard/chainguard-images/vuln-comparison/) section IMO and then you could link to that from here.

File: content/chainguard/chainguard-images/getting-started/nemo/index.md [hunk 4]
```diff
@@ -0,0 +1,3013 @@
...
+      <td>GHSA-g7vv-2v7x-gj9p</td>
+      <td>Low</td>
+    </tr>
+  </tbody>
+</table>
+{{< /details >}}
+
+
+
+## Notes on the Script
```
Comment (erikaheidi) [inline content/chainguard/chainguard-images/getting-started/nemo/index.md:2987]: I would merge this and the next sections under a `## Final Considerations` header where you give advice on next steps for exploration and resources to learn more.

**Review by erikaheidi — CHANGES_REQUESTED** [review]
Hey Patrick! I had some issues to run the example using my GPU. I am not sure what I did wrong! In any case, I did the editorial review and have some suggestions to improve the flow of the article and make it easier to follow. I will try to debug my issue with Docker for the tech run.

File: content/chainguard/chainguard-images/getting-started/nemo/index.md [hunk 5]
```diff
@@ -0,0 +1,3013 @@
...
+```bash
+docker run -it --rm \
+  --gpus all \
+  --shm-size=8g \
+  --ulimit memlock=-1 \
+  --ulimit stack=67108864 \
+  cgr.dev/chainguard/nemo:latest
+```
+
+These options allow access to connected GPU, allocate more shared memory to the container, and put an upper bound on container memory use.
```
[the remark this answers is not in the document]
Comment (smythp) [reply content/chainguard/chainguard-images/getting-started/nemo/index.md:47]: Changed to make the language more specific as in your example, but kep the parallel structure. :+1:

File: content/chainguard/chainguard-images/getting-started/nemo/index.md [hunk 6]
```diff
@@ -0,0 +1,3013 @@
...
+  --gpus all \
+  --shm-size=8g \
+  --ulimit memlock=-1 \
+  --ulimit stack=67108864 \
+  cgr.dev/chainguard/nemo:latest
+```
+
+These options allow access to connected GPU, allocate more shared memory to the container, and put an upper bound on container memory use. 
+
+Running the above for the first time may take a few minutes to pull the NeMo Chainguard Image, currently 4.05 GB. Once the image runs, you will be interacting with a bash shell in the running container. Enter the following commands at the prompt to check the availability of your GPU.
```
[the remark this answers is not in the document]
Comment (smythp) [reply content/chainguard/chainguard-images/getting-started/nemo/index.md:49]: I agree that it would be good to include this, though getting CUDA to work on a local machine is (very) nontrivial. Mac OS is not supported at all if that's the OS you're using. I've written a new blurb that will be included in a prerequisites section. It links to guides on setting  up CUDA for Linux, Windows, GCP, and AWS. As a practical matter I recommend testing on GCP, it's pretty uphill getting this working without preselecting your hardware, OS, and so on for this task. I have updated [this page](https://github.com/chainguard-images/images/blob/main/images/pytorch-cuda12/TESTING.md) with instructions for setting up GCP.

If you were using Mac OS, it looks like that isn't supported at all. :(

Here's the blurb:

<p>
  <a href="https://developer.nvidia.com/cuda-toolkit">Compute Unified Device Architecture (CUDA)</a> is a parallel computing platform developed by NVIDIA. To take advantage of connected NVIDIA GPUs, you’ll need to follow the setup instructions for your local machine or create a CUDA-enabled instance on a cloud provider.
</p>
<p>
  To set up CUDA on your local machine, follow the installation instructions for <a href="https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html">Linux</a> or <a href="https://docs.nvidia.com/cuda/cuda-installation-guide-microsoft-windows/index.html">Windows</a>. CUDA is not currently supported on Mac OS.
</p>
<p>
  Google Cloud Platform provides CUDA-ready <a href="https://cloud.google.com/deep-learning-vm">deep learning instances</a>, including <a href="https://cloud.google.com/deep-learning-vm/docs/pytorch_start_instance">PyTorch-specific instructions</a>. Amazon Web Services also provides CUDA-ready <a href="https://aws.amazon.com/machine-learning/amis/">deep learning instances</a>.
</p>

File: content/chainguard/chainguard-images/getting-started/nemo/index.md (hunk 3 above)
Comment (smythp) [reply content/chainguard/chainguard-images/getting-started/nemo/index.md:142]: A version of this is going out to beta clients and they want this in for now. I'll look into (re)moving this when things settle down.

File: content/chainguard/chainguard-images/getting-started/nemo/index.md [hunk 7]
```diff
@@ -0,0 +1,3013 @@
...
+{{< blurb/deep-learning >}}
+{{< /details >}}
+
+In this getting started guide, we will use the NeMo Chainguard Image to generate speech from plain text using models provided by NeMo's text-to-speech (TTS) and natural language processing (NLP) collections. In doing so, we'll compare the security and footprint of the NeMo Chainguard Image to the official runtime image distributed by NVIDIA and consider further approaches and resources for applying the NeMo Chainguard Image to additional tasks in conversational AI.
+
+This guide is primarily designed for use in an environment with access to one or more NVIDIA GPUs. However, NVIDIA NeMo is built on [PyTorch Lightning](https://lightning.ai/docs/pytorch/stable/), which supports a wide variety of [accelerators](https://pytorch-lightning.readthedocs.io/en/1.1.8/accelerators.html), or interfaces to categories of processing units (CPU, GPU, TPU) or high-level clustering mechanisms such as [Distributed Data Parallel](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html). Some consideration will be given to alternative computing environments such as CPU in this tutorial.
+
+## Testing Access to GPUs
```
[the remark this answers is not in the document]
Comment (smythp) [reply content/chainguard/chainguard-images/getting-started/nemo/index.md:30]: I don't think this tutorial lends itself to steps in the headings. However, given that, the language (first step) in that first paragraph is confusing. I've edited it to remove that language, thanks!

File: content/chainguard/chainguard-images/getting-started/nemo/index.md [hunk 8]
```diff
@@ -0,0 +1,3013 @@
...
+Next, let's download our [tts.py](https://github.com/chainguard-dev/nemo-examples/blob/main/tts.py) script:
+
+```bash
+curl https://raw.githubusercontent.com/chainguard-dev/nemo-examples/main/tts.py > tts.py
+```
+
+You should now be in a working directory with `tts.py`.
+
+```bash
+$ ls
```
[the remark this answers is not in the document]
Comment (smythp) [reply content/chainguard/chainguard-images/getting-started/nemo/index.md:105]: I've removed the `ls` section and added in a little more language in the prose to get across the same thing.

File: content/chainguard/chainguard-images/getting-started/nemo/index.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (smythp) [reply content/chainguard/chainguard-images/getting-started/nemo/index.md:132]: I removed one `ls` command and left the other, as I feel like it shows the final state more clearly than another paragraph, etc. I hear you on the $, I removed them from the docker commands as they do make copying a pain. We might want to talk as a team about options for including prompts in a non-copyable way on CG Academy, something like a smart shortcode or similar.

File: content/chainguard/chainguard-images/getting-started/nemo/index.md (hunk 4 above)
Comment (smythp) [reply content/chainguard/chainguard-images/getting-started/nemo/index.md:2987]: I took out the script review section and added the content to other sections or removed it. :+1:

File: content/chainguard/chainguard-images/getting-started/nemo/index.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (erikaheidi) [reply content/chainguard/chainguard-images/getting-started/nemo/index.md]: Thank you for adding this section! As someone totally newbie to this area, I did not know things wouldn't work out of the box, so I think it's better to make sure users are aware of everything needed to test the image. I am on Linux, going now through the installation instructions to try to run it locally.

**Review by erikaheidi — APPROVED** [review]
I created a VM on GCloud and tested everything, works like a charm, and it's super cool. I wasn't able to run it locally because of my Ubuntu version (23.10, tried to install the kit for 20.4 but it didn't work).

File: content/chainguard/chainguard-images/getting-started/nemo/index.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (erikaheidi) [reply content/chainguard/chainguard-images/getting-started/nemo/index.md]: I was able to run it on GCP 🎉

**Comment by smythp** [comment]
Awesome, thanks Erika, and thanks for the editorial review!

@ltagliaferri, what do you think of taking out  the image comparison section for now and publishing? We could also wait a couple days for CVE mitigation, there's an active effort.
